In [94]:
import pandas as pd

In [95]:
data = pd.read_csv("/home/datauser/Projects/diabetes_project/data/raw/diabetes.csv")
data.head()

/tmp/ipykernel_198214/445365814.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/home/datauser/Projects/diabetes_project/data/raw/diabetes.csv")


,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,payer_code,medical_specialty,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,Caucasian,Female,[0-10),NaN,6,25,1,1,NaN,Pediatrics-Endocrinology,...,No,No,No,No,No,No,No,No,No,NO
1,Caucasian,Female,[10-20),NaN,1,1,7,3,NaN,NaN,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,AfricanAmerican,Female,[20-30),NaN,1,1,7,2,NaN,NaN,...,No,No,No,No,No,No,No,No,Yes,NO
3,Caucasian,Male,[30-40),NaN,1,1,7,2,NaN,NaN,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,Caucasian,Male,[40-50),NaN,1,1,7,1,NaN,NaN,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [96]:
data.shape

(101766, 48)

In [97]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 48 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   race                      99493 non-null   object
 1   gender                    101766 non-null  object
 2   age                       101766 non-null  object
 3   weight                    3197 non-null    object
 4   admission_type_id         101766 non-null  int64 
 5   discharge_disposition_id  101766 non-null  int64 
 6   admission_source_id       101766 non-null  int64 
 7   time_in_hospital          101766 non-null  int64 
 8   payer_code                61510 non-null   object
 9   medical_specialty         51817 non-null   object
 10  num_lab_procedures        101766 non-null  int64 
 11  num_procedures            101766 non-null  int64 
 12  num_medications           101766 non-null  int64 
 13  number_outpatient         101766 non-null  int64 
 14  numb

In [98]:
# Detect and handle missing values
data.isnull().sum()

race                         2273
gender                          0
age                             0
weight                      98569
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                  40256
medical_specialty           49949
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         21
diag_2                        358
diag_3                       1423
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide                   0
glipizide                       0
glyburide     

In [99]:
# Determine percent of missing values in each column
missing_percent = data.isnull().mean() * 100
for col, val in missing_percent.items():
    print(f"{col}: {val:.2f}%")

race: 2.23%
gender: 0.00%
age: 0.00%
weight: 96.86%
admission_type_id: 0.00%
discharge_disposition_id: 0.00%
admission_source_id: 0.00%
time_in_hospital: 0.00%
payer_code: 39.56%
medical_specialty: 49.08%
num_lab_procedures: 0.00%
num_procedures: 0.00%
num_medications: 0.00%
number_outpatient: 0.00%
number_emergency: 0.00%
number_inpatient: 0.00%
diag_1: 0.02%
diag_2: 0.35%
diag_3: 1.40%
number_diagnoses: 0.00%
max_glu_serum: 94.75%
A1Cresult: 83.28%
metformin: 0.00%
repaglinide: 0.00%
nateglinide: 0.00%
chlorpropamide: 0.00%
glimepiride: 0.00%
acetohexamide: 0.00%
glipizide: 0.00%
glyburide: 0.00%
tolbutamide: 0.00%
pioglitazone: 0.00%
rosiglitazone: 0.00%
acarbose: 0.00%
miglitol: 0.00%
troglitazone: 0.00%
tolazamide: 0.00%
examide: 0.00%
citoglipton: 0.00%
insulin: 0.00%
glyburide-metformin: 0.00%
glipizide-metformin: 0.00%
glimepiride-pioglitazone: 0.00%
metformin-rosiglitazone: 0.00%
metformin-pioglitazone: 0.00%
change: 0.00%
diabetesMed: 0.00%
readmitted: 0.00%


### 📊 Résumé des valeurs manquantes

- **Dimensions des données :** `101,766` lignes × `48` colonnes  
- **Cellules manquantes :** `374,017` (≈ **7.66%** du total)  
- **Colonnes avec des valeurs manquantes :** `9 / 48`  
- **Colonnes sans valeurs manquantes :** `39 / 48`  
- **Colonne cible (`readmitted`) :** `0.00%` de valeurs manquantes ✅  

#### 🛑 Colonnes les plus affectées (Top 5) :

| Colonne             | % de valeurs manquantes |
|---------------------|--------------------------|
| `weight`            | 96.86%                   |
| `max_glu_serum`     | 94.75%                   |
| `A1Cresult`         | 83.28%                   |
| `medical_specialty` | 49.08%                   |
| `payer_code`        | 39.56%                   |

In [100]:
# Supprimer les colonnes dont le % de valeurs manquantes est supérieur à un seuil
threshold = 80.0  # Seuil en pourcentage
cols_to_drop = [col for col, val in missing_percent.items() if val > threshold]
print(f"Colonnes supprimées : {cols_to_drop}")

data.drop(columns=cols_to_drop, inplace=True)

Colonnes supprimées : ['weight', 'max_glu_serum', 'A1Cresult']


In [101]:
data_object_with_NAN = ['race', 'payer_code', 'medical_specialty', 'diag_1', 'diag_2', 'diag_3']
for col in data_object_with_NAN:
    data[col] = data[col].fillna('Unknown')

In [102]:
# Detect and handle missing values
data.isnull().sum()

race                        0
gender                      0
age                         0
admission_type_id           0
discharge_disposition_id    0
admission_source_id         0
time_in_hospital            0
payer_code                  0
medical_specialty           0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
number_diagnoses            0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide

In [103]:
# Recuperer toutes les colones en object
object_cols = data.select_dtypes(include=['object']).columns
for col in object_cols:
    data[col] = data[col].str.strip().str.lower().str.replace(' ', '_')

In [104]:
data.head()

,race,gender,age,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,payer_code,medical_specialty,num_lab_procedures,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,caucasian,female,[0-10),6,25,1,1,unknown,pediatrics-endocrinology,41,...,no,no,no,no,no,no,no,no,no,no
1,caucasian,female,[10-20),1,1,7,3,unknown,unknown,59,...,no,up,no,no,no,no,no,ch,yes,>30
2,africanamerican,female,[20-30),1,1,7,2,unknown,unknown,11,...,no,no,no,no,no,no,no,no,yes,no
3,caucasian,male,[30-40),1,1,7,2,unknown,unknown,44,...,no,up,no,no,no,no,no,ch,yes,no
4,caucasian,male,[40-50),1,1,7,1,unknown,unknown,51,...,no,steady,no,no,no,no,no,ch,yes,no


In [105]:
# Voir les valeurs uniques de race
data["race"].unique()

array(['caucasian', 'africanamerican', 'unknown', 'other', 'asian',
       'hispanic'], dtype=object)

In [106]:
data['race'] = data['race'].replace({
    'africanamerican': 'african american',
    'unknown': 'unknown',
    'other': 'other',
    'caucasian': 'caucasian',
    'asian': 'asian',
    'hispanic': 'hispanic'
})

In [107]:
# Voir les valeurs uniques de race
data["race"].unique()

array(['caucasian', 'african american', 'unknown', 'other', 'asian',
       'hispanic'], dtype=object)

In [108]:
import random, re

In [109]:
def random_age_from_interval(interval):
    match = re.match(r'\[(\d+)-(\d+)\)', interval)
    if match:
        start = int(match.group(1))
        end = int(match.group(2))
        return random.randint(start, end - 1)
    else:
        return None

In [110]:
data['age_num'] = data['age'].apply(random_age_from_interval)

In [111]:
for i, col in enumerate(data.columns):
    print(i, col)


0 race
1 gender
2 age
3 admission_type_id
4 discharge_disposition_id
5 admission_source_id
6 time_in_hospital
7 payer_code
8 medical_specialty
9 num_lab_procedures
10 num_procedures
11 num_medications
12 number_outpatient
13 number_emergency
14 number_inpatient
15 diag_1
16 diag_2
17 diag_3
18 number_diagnoses
19 metformin
20 repaglinide
21 nateglinide
22 chlorpropamide
23 glimepiride
24 acetohexamide
25 glipizide
26 glyburide
27 tolbutamide
28 pioglitazone
29 rosiglitazone
30 acarbose
31 miglitol
32 troglitazone
33 tolazamide
34 examide
35 citoglipton
36 insulin
37 glyburide-metformin
38 glipizide-metformin
39 glimepiride-pioglitazone
40 metformin-rosiglitazone
41 metformin-pioglitazone
42 change
43 diabetesMed
44 readmitted
45 age_num


In [112]:
# data['medical_specialty'] = data['medical_specialty'].replace(
#     'obsterics&gynecology-gynecologiconco', 
#     'obstetrics-gynecology-gynecologic-oncology'
# )
# data['medical_specialty'] = data['medical_specialty'].replace(
#     'surgery-plasticwithinheadandneck',
#     'surgery-plastic-within-head-and-neck'
# )

# data['medical_specialty'] = data['medical_specialty'].str.replace('/', '-', regex=False)
# data['medical_specialty'] = data['medical_specialty'].str.replace('&', '-', regex=False)


In [113]:
corrections = {
    'obstetricsandgynecology': 'obstetrics-gynecology',
    'surgery-plasticwithinheadandneck': 'surgery-plastic-within-head-and-neck',
    'pediatrics-emergencymedicine': 'pediatrics-emergency-medicine',
    'physicalmedicineandrehabilitation': 'physical-medicine-and-rehabilitation',
    'allergyandimmunology': 'allergy-and-immunology',
    'pediatrics-infectiousdiseases': 'pediatrics-infectious-diseases',
    'pediatrics-allergyandimmunology': 'pediatrics-allergy-and-immunology',
    'physiciannotfound': "unknown",  
    'surgicalspecialty': "unknown",  
    'dcpteam': "unknown",
    'resident': "unknown",
    'surgeon': "unknown",
    'outreachservices': "unknown",
    'speech': 'speech-therapy',
    'sportsmedicine': 'sports-medicine',
    'hematology': 'hematology-oncology',
    'radiologist': 'radiology',
    'obstetrics': 'obstetrics-gynecology',
    'psychiatry-addictive': 'psychiatry-addiction-medicine',
}

data['medical_specialty'] = data['medical_specialty'].str.lower()  # uniformiser en minuscules
data['medical_specialty'] = data['medical_specialty'].str.replace('/', '-', regex=False)
data['medical_specialty'] = data['medical_specialty'].str.replace('&', '-', regex=False)
data['medical_specialty'] = data['medical_specialty'].str.replace(' ', '-')  # optionnel, si besoin

# Appliquer les corrections
data['medical_specialty'] = data['medical_specialty'].replace(corrections)

# Optionnel : remplacer les valeurs inconnues ou manquantes par 'unknown'
data['medical_specialty'] = data['medical_specialty'].fillna('unknown')

# Vérifier le résultat
print(data['medical_specialty'].unique())

['pediatrics-endocrinology' 'unknown' 'internalmedicine'
 'family-generalpractice' 'cardiology' 'surgery-general' 'orthopedics'
 'gastroenterology' 'surgery-cardiovascular-thoracic' 'nephrology'
 'orthopedics-reconstructive' 'psychiatry' 'emergency-trauma'
 'pulmonology' 'surgery-neuro' 'obsterics-gynecology-gynecologiconco'
 'obstetrics-gynecology' 'pediatrics' 'hematology-oncology'
 'otolaryngology' 'surgery-colon-rectal' 'pediatrics-criticalcare'
 'endocrinology' 'urology' 'psychiatry-child-adolescent'
 'pediatrics-pulmonology' 'neurology' 'anesthesiology-pediatric'
 'radiology' 'pediatrics-hematology-oncology' 'psychology' 'podiatry'
 'gynecology' 'oncology' 'pediatrics-neurology' 'surgery-plastic'
 'surgery-thoracic' 'surgery-plastic-within-head-and-neck' 'ophthalmology'
 'surgery-pediatric' 'pediatrics-emergency-medicine'
 'physical-medicine-and-rehabilitation' 'infectiousdiseases'
 'anesthesiology' 'rheumatology' 'allergy-and-immunology'
 'surgery-maxillofacial' 'pediatrics-infe